### PR TITLE
Feat: Milestone payload options

### PR DESCRIPTION
**Pull request by brancoder**
# Description of change

Update milestone payload to include milestone options in complience with tip-29:

Tip-29: 
https://github.com/iotaledger/tips/commit/8e4642b235d02d7c58b039524d1b55c73612e2c8

Fixes: #837 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/iota/src/binary/milestoneOptions/milestoneOptions.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+);
+
+
+/**
+ * Deserialize the milestone options from binary.
+ * @param readStream The stream to read the data from.
+ * @returns The deserialized object.
+ */
+ export function deserializeMilestoneOptions(readStream: ReadStream): MilestoneOptionTypes[] {
+    const numMilestoneOptions = readStream.readUInt8("milestoneOptions.numMilestoneOptions");
```
Comment (msarcev) [inline packages/iota/src/binary/milestoneOptions/milestoneOptions.ts]: Should we have the same naming here as in the serialize side on line 50 ? And maybe name the constant like that also ? 
i.e. `const optionsCount = readStream.readUInt8("milestoneOptions.optionsCount");`

File: packages/iota/src/binary/milestoneOptions/receiptMilestoneOption.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const type = readStream.readUInt8("receiptMilestoneOption.type");
+    if (type !== RECEIPT_MILESTONE_OPTION_TYPE) {
+        throw new Error(`Type mismatch in receiptMilestoneOption ${type}`);
+    }
+    const migratedAt = readStream.readUInt32("receiptMilestoneOption.migratedAt");
+    const final = readStream.readBoolean("receiptMilestoneOption.final");
+
+    const funds = deserializeFunds(readStream);
+    const treasuryTransactionPayload = deserializePayload(readStream);
+    if (!treasuryTransactionPayload || treasuryTransactionPayload.type !== TREASURY_TRANSACTION_PAYLOAD_TYPE) {
```
Comment (msarcev) [inline packages/iota/src/binary/milestoneOptions/receiptMilestoneOption.ts]: Here we could simplify the condition with optional chaining.
i.e.`if (treasuryTransactionPayload?.type !== TREASURY_TRANSACTION_PAYLOAD_TYPE) {`

File: packages/iota/src/models/payloads/IMilestonePayload.ts [hunk 3]
```diff
@@ -24,37 +24,37 @@ export interface IMilestonePayload extends ITypeBase<7> {
...
     /**
-     * The merkle proof inclusions.
+     * The parents where this milestone attaches to.
      */
-    inclusionMerkleProof: string;
+    parentMessageIds: string[];
 
     /**
-     * The next PoW score.
+     * The merkle proof confirmation.
```
Comment (msarcev) [inline packages/iota/src/models/payloads/IMilestonePayload.ts]: Here we could give a better description taken from TIP-29, i.e. 
`The Merkle tree hash of the Message IDs of all messages confirmed by this milestone.` or 
`The Merkle tree hash of all messages confirmed by this milestone.`

File: packages/iota/src/models/payloads/IMilestonePayload.ts [hunk 4]
```diff
@@ -24,37 +24,37 @@ export interface IMilestonePayload extends ITypeBase<7> {
...
      */
-    nextPoWScore: number;
+    confirmedMerkleRoot: string;
 
     /**
-     * The milestone at which the next PoW score becomes active.
+     * The merkle proof applied.
```
Comment (msarcev) [inline packages/iota/src/models/payloads/IMilestonePayload.ts]: Also here description from TIP sounds better. i.e.
`The Merkle tree hash of all the messages applied by this milestone that contain a state-mutating transaction.`

**Review by msarcev — APPROVED** [review]
A couple of minor comments, other than that LGTM.

File: packages/iota/src/models/milestoneOptions/IPoWMilestoneOption.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+// SPDX-License-Identifier: Apache-2.0
+import type { ITypeBase } from "../ITypeBase";
+
+/**
+ * The global type for the payload.
+ */
+export const POW_MILESTONE_OPTION_TYPE = 1;
+
+/**
+ * Receipt payload.
```
Comment (laumair) [inline packages/iota/src/models/milestoneOptions/IPoWMilestoneOption.ts]: @brancoder guess this is a typo.

**Review by laumair — APPROVED** [review]
Looks good overall. 

One small thing is that I have noticed that we are not checking (package) lock files with any package but for `iota/` package in this PR, we are adding the lock file. Probably best to remove it for consistency.

**Comment by laumair** [comment]
We should also probably get a review from @Thoralf-M and / or @lzpap.

File: packages/iota/test/binary/payloads/milestonePayload.spec.ts [hunk 6]
```diff
@@ -4,25 +4,28 @@ import { Converter, ReadStream, WriteStream } from "@iota/util.js";
...
                 "0x04ba147c9cc9bebd3b97310a23d385f33d8e67ac42868b69bc06f5468e3c0a02",
                 "0xc0ab1d1f6886ba6317634da6b2d957e7c987a9699dd3707d1e2751fcf4b8efe3"
             ],
-            inclusionMerkleProof: "0x786a02f742015903c6c6fd852552d272912f4740e15847618a86e217f71f5419",
-            nextPoWScore: 0,
-            nextPoWScoreMilestoneIndex: 1,
+            confirmedMerkleRoot: "0x665de8d34bca02af275a6ccaf2d5f7b1d018f695473f19855d7ad1a54f106ed1",
+            appliedMerkleRoot: "0x0e5751c026e543b2e8ab2eb06099daa1d1e5df47778f7787faab45cdf12fe3a8",
             metadata: "0x1111111122222222",
+            // options:[],
```
Comment (Thoralf-M) [inline packages/iota/test/binary/payloads/milestonePayload.spec.ts]: Any reason for this comment?

File: packages/iota/src/binary/milestoneOptions/milestoneOptions.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 IOTA Stiftung
```
Comment (Thoralf-M) [inline packages/iota/src/binary/milestoneOptions/milestoneOptions.ts:1]: We're already in 2022, also in some other files

Comment (brancoder) [reply packages/iota/src/binary/milestoneOptions/milestoneOptions.ts:1]: I will create separate PR to update the Copyright in the entire project.

**Review by lzpap — APPROVED** [review]

**Pull request merged**